kani-compiler/src/kani_middle/transform/loop_contracts.rs

```diff
@@ -0,0 +1,467 @@
+///    ...
+///    loop_body_blocks
+///    ...
+///
+///    loop_latch_block: {
+///         loop_latch_stmts
+///         goto -> bb_idx;
+///    }
+///    ```
+///    to blocks
+///    ```ignore
+///    bb_idx: {
+///         _v = true
+///         goto -> terminator_target
+///    }
+///
+///    ...
+///    loop_body_blocks
+///    ...
+///
+///    loop_latch_block: {
+///         loop_latch_stmts
+///         goto -> bb_new_loop_latch;
+///    }
+///
+///    bb_new_loop_latch: {
+///         loop_head_body
+///         _v = kani_register_loop_contract(move args) -> [return: bb_idx];
+///    }
+///    ```
+///
+/// 3. Move the statements `loop_head_stmts` of loop head  with contracts to the body of register
+///    functions for later codegen them as statement_expression in CBMC loop contracts.
+#[derive(Debug, Default)]
+pub struct LoopContractPass {
+    /// Cache KaniRunContract function used to implement contracts.
+    run_contract_fn: Option<FnDef>,
+    /// The bb_idx of the new loop latch block.
+    /// Keys are the original loop head.
+    new_loop_latches: HashMap<usize, usize>,
+    /// Statements of loop head with loop contracts.
+    registered_stmts: HashMap<DefId, Vec<Statement>>,
+    /// If loop contracts is enabled.
+    loop_contracts_enabled: bool,
+}
+
+impl GlobalPass for LoopContractPass {
+    fn is_enabled(&self, _query_db: &QueryDb) -> bool
+    where
+        Self: Sized,
+    {
+        true
+    }
+
+    /// Run a transformation pass on the whole codegen unit.
+    fn transform(
+        &mut self,
+        tcx: TyCtxt,
+        _call_graph: &CallGraph,
+        _starting_items: &[MonoItem],
+        instances: Vec<Instance>,
+        transformer: &mut BodyTransformation,
+    ) {
+        if self.loop_contracts_enabled {
+            // First transform functions with loop contracts.
+            for instance in &instances {
+                let body = instance.body().unwrap();
+                let (modified, new_body) = self.transform_main_body(tcx, body, *instance);
+                if modified {
+                    transformer.cache.entry(*instance).and_modify(|transformation_result| {
+                        *transformation_result = TransformationResult::Modified(new_body);
+                    });
+                }
+            }
+
+            // Move loop head statements with loop contracts to the corresponding register functions.
+            for instance in &instances {
+                let body = instance.body().unwrap();
+                let (modified, new_body) = self.transform_register_body(tcx, body, *instance);
+                if modified {
+                    transformer.cache.entry(*instance).and_modify(|transformation_result| {
+                        *transformation_result = TransformationResult::Modified(new_body);
+                    });
+                }
+            }
+        } else {
+            for instance in &instances {
+                let body = instance.body().unwrap();
```

Comment:
We shouldn't be using `instance.body()` since they don't contain changes from other transformation passes. This should also go away by using `TransformPass`.